File: taichi/program/graph.cpp
#include "taichi/program/graph.h"
#include "taichi/program/kernel.h"
#include "taichi/aot/module_builder.h"
#include "spdlog/fmt/fmt.h"

#include <fstream>

namespace taichi {
namespace lang {

void Dispatch::compile(std::vector<aot::CompiledDispatch> &dispatches) {
  compiled_kernel_ = kernel_->compile_to_aot_kernel();
  aot::CompiledDispatch dispatch{kernel_->get_name(), symbolic_args_,
                                 compiled_kernel_.get()};
  dispatches.push_back(std::move(dispatch));
}

void Sequential::compile(std::vector<aot::CompiledDispatch> &dispatches) {
  // In the future we can do more across-kernel optimization here.
  for (Node *n : sequence_) {
    n->compile(dispatches);
  }
}

void Sequential::append(Node *node) {
  sequence_.push_back(node);
}

void Sequential::emplace(Kernel *kernel, const std::vector<Arg> &args) {
  Node *n = owning_graph_->create_dispatch(kernel, args);
  sequence_.push_back(n);
}

Graph::Graph(std::string name) : name_(name) {
  seq_ = std::make_unique<Sequential>(this);
}
Node *Graph::create_dispatch(Kernel *kernel, const std::vector<Arg> &args) {
  all_nodes_.push_back(std::make_unique<Dispatch>(kernel, args));
  return all_nodes_.back().get();
}

Sequential *Graph::create_sequential() {
  all_nodes_.push_back(std::make_unique<Sequential>(this));
  auto res = static_cast<Sequential *>(all_nodes_.back().get());

  return res;
}

void Graph::compile() {
  seq()->compile(compiled_graph_.dispatches);
}

Sequential *Graph::seq() const {
  return seq_.get();
}

void Graph::emplace(Kernel *kernel, const std::vector<Arg> &args) {
  seq()->emplace(kernel, args);
}

void Graph::run(std::unordered_map<std::string, IValue> &args) {
  RuntimeContext ctx;
  for (const auto &dispatch : compiled_graph_.dispatches) {
    memset(&ctx, 0, sizeof(RuntimeContext));
    auto &symbolic_args_ = dispatch.symbolic_args;
    for (int i = 0; i < symbolic_args_.size(); ++i) {
      auto &symbolic_arg = symbolic_args_[i];
      auto found = args.find(symbolic_arg.name);
      TI_ERROR_IF(found == args.end(), "Missing runtime value for ",
                  symbolic_arg.name);
      IValue &ival = found->second;
      if (ival.tag == ArgKind::NDARRAY) {
        Ndarray *arr = reinterpret_cast<Ndarray *>(ival.val);
        TI_ERROR_IF((symbolic_arg.tag != ival.tag) ||
                        (symbolic_arg.element_shape != arr->shape),
                    "Mismatched shape information for argument",
                    symbolic_arg.name);
        set_runtime_ctx_ndarray(&ctx, i, arr);
      } else {
        TI_ERROR_IF(symbolic_arg.tag != ArgKind::SCALAR,
                    "Required a scalar for argument", symbolic_arg.name);
        ctx.set_arg(i, ival.val);
      }
    }

    dispatch.compiled_kernel->launch(&ctx);
  }
}
}  // namespace lang
}  // namespace taichi

File: python/taichi/graph/_graph.py
from taichi._lib import core as _ti_core
from taichi.aot.utils import produce_injected_args
from taichi.lang import kernel_impl

ArgKind = _ti_core.ArgKind
Arg = _ti_core.Arg


class Sequential:
    def __init__(self, seq):
        self.seq_ = seq

    def emplace(self, kernel_fn, *args):
        kernel = kernel_fn._primal
        assert isinstance(kernel, kernel_impl.Kernel)
        injected_args = produce_injected_args(kernel,
                                              template_args=None,
                                              symbolic_args=args)
        # FIXME: move compilation down to the compile() below
        kernel.ensure_compiled(*injected_args)
        self.seq_.emplace(kernel.kernel_cpp, args)


class Graph:
    def __init__(self, name):
        self._graph = _ti_core.Graph(name)

    def emplace(self, kernel_fn, *args):
        kernel = kernel_fn._primal
        assert isinstance(kernel, kernel_impl.Kernel)
        injected_args = produce_injected_args(kernel,
                                              template_args=None,
                                              symbolic_args=args)
        kernel.ensure_compiled(*injected_args)
        self._graph.emplace(kernel.kernel_cpp, args)

    def create_sequential(self):
        return Sequential(self._graph.create_sequential())

    def append(self, node):
        # FIXME: support this in a cleaner way
        if isinstance(node, Sequential):
            self._graph.seq().append(node.seq_)

    def compile(self):
        self._graph.compile()

    def run(self, args):
        arg_ptrs = {k: args[k].arr for k in args.keys()}
        self._graph.run(arg_ptrs)


__all__ = ['Graph', 'Arg', 'ArgKind']
